Port LIS2MDL driver on top of stmemsc i/f

avisconti opened the pull request:
Port the lis2mdl sensor driver on top of the lis2mdl_StdC HAL interface (in modules/hal/st/sensor/stmemsc/). The new driver is using I2C/SPI bus.

This PR is also modifing x_nucleo_iks01a3 board and sample for a proper testing of the
driver in I2C mode.

avisconti committed:
board/shields: x-nucleo-iks01a3: configure irq-gpios for lis2mdl

Configure interrupt gpio information (irq-gpios) into the
x-nucleo-iks01a3 shield overlay file.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
sample/shield: x-nucleo-iks01a3: add trig/temp/cfg support to LIS2MDL

Add support to LIS2MDL trigger, dynamic odr configuration as well as
temperature data reading.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

dbkinder reviewed (approved):
doc changes LGTM, thanks!

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -1,9 +1,11 @@
+/* ST Microelectronics LIS2MDL 3-axis magnetometer sensor
  *
- * LIS2MDL mag driver
+ * Copyright (c) 2019 STMicroelectronics
```
Change the year to 2018-2019

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -134,7 +94,7 @@ static void lis2mdl_channel_get_mag(struct device *dev,
 
 	for (i = ofs_start; i <= ofs_stop; i++) {
-		cval = lis2mdl->mag[i] * lis2mdl->mag_fs_sensitivity;
+		cval = lis2mdl->mag[i] * 1500;
```
Is this change really related to stmemsc i/f?

avisconti replied:
No, it is a code cleaning. The sensitivity does not change and it is always 1500 mgauss/LSB.
I would keep it in the same commit even if not strictly related to stmemsc. Do you agree?

MaureenHelm replied:
No, please move to a separate commit

avisconti replied:
ok, done!

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -148,7 +108,7 @@ static void lis2mdl_channel_get_temp(struct device *dev,
 
 	val->val1 = drv_data->temp_sample / 100;
-	val->val2 = drv_data->temp_sample % 100;
+	val->val2 = (drv_data->temp_sample % 100) * 10000;
```
Is this change really related to stmemsc i/f?

avisconti replied:
No, I realized this was a bug. I can put it in a separate commit

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.h:
```diff
@@ -1,114 +1,54 @@
+/* ST Microelectronics LIS2MDL 3-axis magnetometer sensor
  *
- * LIS2MDL mag header file
+ * Copyright (c) 2019 STMicroelectronics
```
Change the year to 2018-2019

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl.h:
```diff
@@ -1,114 +1,54 @@
 
+#include <drivers/spi.h>
 #include <drivers/gpio.h>
+#include <sensor.h>
```
drivers/sensor.h

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl_trigger.c:
```diff
@@ -1,9 +1,11 @@
+/* ST Microelectronics LIS2MDL 3-axis magnetometer sensor
  *
- * LIS2MDL mag interrupt configuration and management
+ * Copyright (c) 2019 STMicroelectronics
```
Change the year to 2018-2019

MaureenHelm commented on drivers/sensor/lis2mdl/lis2mdl_trigger.c:
```diff
@@ -155,7 +135,18 @@ int lis2mdl_init_interrupt(struct device *dev)
 #endif
 
-	gpio_pin_enable_callback(lis2mdl->gpio, config->gpio_pin);
+	gpio_pin_configure(lis2mdl->gpio, config->gpio_pin,
```
Is this change really related to stmemsc i/f?

avisconti replied:
Not related to stmemsc. I thought it was good to re-write it in the way all the other drivers, at least the ST ones, are done. Moreover, I think that it is better to create the callback thread prior to enable the callback.

avisconti replied:
I would keep it in the same commit even if not strictly related to stmemsc. Do you agree?

MaureenHelm replied:
It's fine to keep in the same PR, but I'd prefer it to be in a separate commit if it's not related to stmemsc

avisconti replied:
ACK.
Done in this way and re-pushed.

avisconti committed:
driver/sensor: lis2mdl: Fix the temperature fractional value

The temperature fractional value has to be assigned multiplied
by 10^6.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@MaureenHelm 
I re-pushed it adding your changes except a couple of things highlighted above. Please let me know if you agree.

Moreover, I noticed that I made a type concerning CS_GPIOS DT instance, found just because I built it for SPI as well. So I also made that change.

avisconti committed:
driver/sensor: lis2mdl: change order of operations in trigger init

Creation of callback thread is moved before the enabling of
the callback itself.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
driver/sensor: lis2mdl: make use of STdC definitions

Port the lis2mdl sensor driver on top of the lis2mdl_StdC
HAL interface (in modules/hal/st/sensor/stmemsc/).

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
driver/sensor: lis2mdl: remove sensitivity

LIS2MDL sensor has a fixed sensitivity equal to 1500 uGauss/LSB.
So, use a constant value directly.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@MaureenHelm 
Every point should be addressed now.
